File: example-1/graph.cpp
#include <iostream>
using namespace std;

struct Node {
    int value;     
    Node* next;     
};

struct List {
    Node* head;     
};

class Graph {
private:
    List* array;    
    int numVertices;
    
public:
    Graph(int vertices) {
        numVertices = vertices;
        array = new List[vertices];
        for (int i = 0; i < vertices; i++) {
            array[i].head = NULL;
        }
    }
    
    void addVertex(int v) {
        if (v >= numVertices) {
            List* newArray = new List[v + 1];
            for (int i = 0; i < numVertices; i++) {
                newArray[i] = array[i];
            }
            
            for (int i = numVertices; i <= v; i++) {
                newArray[i].head = NULL;
            }
            
            delete[] array;
            array = newArray;
            numVertices = v + 1;
        }
    }
    
    void addEdge(int v1, int v2) {
        if (v1 >= numVertices) addVertex(v1);
        if (v2 >= numVertices) addVertex(v2);
        
        Node* newNode = new Node;
        newNode->value = v2;
        newNode->next = array[v1].head;
        array[v1].head = newNode;
        
        newNode = new Node;
        newNode->value = v1;
        newNode->next = array[v2].head;
        array[v2].head = newNode;
    }
    
    void printGraph() {
        for (int i = 0; i < numVertices; i++) {
            cout << "Vértice " << i << endl;
            Node* temp = array[i].head;
            while (temp) {
                cout << " -> " << temp->value;
                temp = temp->next;
            }
            cout << endl;
        }
    }
};

int main() {
    
    Graph graph(5);
    graph.addEdge(0, 1);
    graph.addEdge(0, 4);
    graph.addEdge(1, 2);
    graph.addEdge(1, 3);
    graph.addEdge(1, 4);
    graph.addEdge(2, 3);
    graph.addEdge(3, 4);
    graph.addEdge(4, 5);
    
    
    cout << "Estrutura do grafo:" << endl;
    graph.printGraph();
    
    return 0;
}

File: example-2/graph.py
class Node:
    def __init__(self, value: str, weight: int = None):
        self.value = value
        self.next = None
        self.weight = weight

class List:
    def __init__(self):
        self.head = None

class Graph:
    def __init__(self, vertices: list[str]):
        self.vertices = vertices
        self.num_vertices = len(vertices)
        self.array = [List() for _ in range(self.num_vertices)]
        self.vertex_index = {v: i for i, v in enumerate(vertices)}
    
    def add_vertex(self, v: str):
        if v not in self.vertex_index:
            self.vertices.append(v)
            self.vertex_index[v] = self.num_vertices
            new_array = [List() for _ in range(self.num_vertices + 1)]
            for i in range(self.num_vertices):
                new_array[i] = self.array[i]
            
            self.array = new_array
            self.num_vertices += 1
    
    def add_edge(self, v1: str, v2: str, weight: int = None):

        if v1 not in self.vertex_index:
            self.add_vertex(v1)
        if v2 not in self.vertex_index:
            self.add_vertex(v2)
        

        i1 = self.vertex_index[v1]
        i2 = self.vertex_index[v2]
        

        new_node = Node(v2, weight)
        new_node.next = self.array[i1].head
        self.array[i1].head = new_node
        

        new_node = Node(v1, weight)
        new_node.next = self.array[i2].head
        self.array[i2].head = new_node
    
    def print_graph(self):
        for v, i in self.vertex_index.items():
            print(f"Vértice {v}")
            temp = self.array[i].head
            while temp:
                print(f" -> {temp.value}|{temp.weight}|", end="")
                temp = temp.next
            print()
